File: Behavioral/Strategy/kotlin/Strategy.kt
class Product(val name: String, val price: Double)

interface BillingStrategy {
    fun getTotal(productList:List<Product>) : Double
}

class NormalStrategy : BillingStrategy {
    override fun getTotal(productList: List<Product>): Double {
        return productList.map { it.price }.sum()
    }
}

class DiscountStrategy(val discount : Double) : BillingStrategy {
    override fun getTotal(productList: List<Product>): Double {
        return productList.map { it.price }.sum() * (1 - discount / 100)
    }

}


fun main(args: Array<String>) {
    val productA = Product("Product A", 10.0)
    val productB = Product("Product B", 25.6)

    val products = listOf<Product>(productA, productB)

    val normalStrategy = NormalStrategy()
    println(normalStrategy.getTotal(products))

    val discountStrategy = DiscountStrategy(10.0)
    println(discountStrategy.getTotal(products))
}

File: Behavioral/Strategy/scala/Strategy.scala
case class Product(name: String, price: Double)

sealed trait BillingStrategy {

  def getTotal(products: List[Product]): Double
}

class NormalStrategy extends BillingStrategy {

  override def getTotal(products: List[Product]): Double = {
    products.map(_.price).sum
  }
}

class DiscountStrategy(discount: Double) extends BillingStrategy {

  override def getTotal(products: List[Product]): Double = {
    products.map(_.price).sum * (1 - discount / 100)
  }
}

object Main extends App {

  def functionalStrategy(products: List[Product], f: (List[Product], Double) => Double, discount: Double = 0): Double = f(products, discount)

  override def main(args: Array[String]): Unit = {
    val productA = Product("Product A", 10.0)
    val productB = Product("Product B", 25.6)
    val products = List(productA, productB)

    val normalStrategy = new NormalStrategy()
    println(normalStrategy.getTotal(products))

    val discountStrategy = new DiscountStrategy(10.0)
    println(discountStrategy.getTotal(products))

    // Functional way of strategy pattern realization
    def normalStrategyFunction(productList: List[Product], discount: Double) = productList.map(_.price).sum
    println(functionalStrategy(products, normalStrategyFunction))

    def discountStrategyFunction(productList: List[Product], discount: Double) = productList.map(_.price).sum * (1 - discount / 100)
    println(functionalStrategy(products, discountStrategyFunction, 10.0))
  }
}
